### ISSUE TITLE
Airgap / Offline Operational Mode

### ISSUE THREAD

**Issue by christophermaier** [habitat-sh/habitat#4039]
Currently, the day-to-day operation of Habitat is rather closely tied to online access to Builder. It would be useful to run Habitat in a kind of "offline" mode, where it will never attempt to go out to Builder, but instead rely solely on locally-available keys and packages.

This can be useful to developers on airplanes, as well as installations that have stringent networking policies / regulations.

I believe many of the pieces we need already exist; we just have to ensure they're assembled in the right (and complete!) way.

**Comment by wduncanfraser** [comment]
Carrying over conversation from Slack, while we do not currently have a need for being fully airgapped, we currently work mostly with internal proprietary hab packages that we build in a pipeline and store in Artifactory.

Our preferred workflow (Which worked until more recent versions of hab) would be to use Chef to download and install hab packages from Artifactory, and then load the service by name, for example: `hab svc load alasconnect/projecta-componentb` and have it use the already installed packages. With recent changes we need to reference the artifact directly, which complicates doing proper guards for seeing if the service is already loaded, etc.

In short, what would be beneficial for us would be when loading a service, being able to flag "Only use installed packages, everything we have is already here", but still allow internet usage for pulling dependencies when we install new versions of our packages.

Let me know what other information would be useful.

**Comment by smartb-pair** [comment]
It feels like a lot of the modifications we would make for Airgapped Mode would also be extremely useful for Dev Studio (specifically, caching not just artifacts but also installed packages).

**Comment by rsertelon** [comment]
Related to this topic - but not sure this is entirely - is the fact that when we export our artifacts to tar, we also need to export hab-launcher to tar and also upload it.

In my mind, tar export is used when no bldr is available , ie: internet access is not available (I can't really see another usage myself).

To help in this matter, it would be nice if the tar export also contained hab-launcher so that untaring the artifact's tar would be the only required step to deploy.

**Comment by tashimi** [comment]
Hmm, `hab-launcher` not being included in a tar export sounds like a bug to me. I'll ask.

**Comment by qhartman** [comment]
I also would have use for an airgapped solution. I was told in slack that using the NO_INSTALL_DEPS flag would mosty-ish enable this if all dependencies were available locally. I haven't been able to confirm or refute this yet, but if true, should address the "dev on an airplane" scenario if their environment were properly provisioned first.

**Comment by christophermaier** [comment]
Some additional information at least tangentially related to this: https://forums.habitat.sh/t/building-software-with-dependencies-not-in-builder/436

**Comment by bdangit** [comment]
As a developer building inside the Studio, we use `INSTALL_NO_DEPS` everytime we build locally as we want to iterate fast on development.  We also notice that public `.harts` do not change that fast, so it’s very nice not having to check the depot for each dependency.

**Comment by elliott-davis** [comment]
Closed by: #4771
